Support for weak build item production and virtual build items

dmlloyd opened the pull request.

dmlloyd committed:
Support for empty build items

dmlloyd committed:
Add produce/consume annotations for build items

dmlloyd committed:
Introduce annotation for weak build step producers

Fixes #2951

emmanuelbernard reviewed (commented):
Hey @dmlloyd 
I don't really grasp in which situations it should be useful and how to use it concretely.
My hunch is that Virtual is a name that is too generic and probably could get a better name.

Also are virtual build items and weak things somewhat related?

At any rate, I would love to see a doc change in the extension guide with concrete examples why an extension would want such tools. That will help me understand better the why, how to use it and ideally propose better names.

dmlloyd commented:
Sure, I can add some docs for these cases.

dmlloyd committed:
Update extension authors guide

Include information on weak and empty build items, as well as other things that have changed since the doc was written

dmlloyd commented:
I've added a commit which includes updated documentation for build steps.

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
-bytecode invocations, which is mapped to a `BuildItem` transparently.
+Work is done at augmentation time by _build steps_ which produce and consume _build items_.  The build steps found in
+the deployment modules that correspond to the extensions in the project build are automatically wired together and executed
+to produce the final build artifact(s).
```
Can you add

Think of build items as a way for different extensions to communicate with one another, for example:

- expose the fact that a database configuration exists
- consume that database configuration (e.g. a connection pool extension or an ORM extension)
- ask an extension to do work for another extension: e.g. an extension wanting to define a new CDI bean and asking the ArC extension

This is a very flexible mechanism.

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
+
+Normally a build step is included whenever it produces a build item which is consumed by any other build step.  In this way,
+only the steps necessary to produce the final artifact(s) are included, and steps which pertain to extensions which are
+not installed or which only produce build items which are not relevant for the given artifact type are excluded.  Sometimes
```
Can you add concrete example of the sometimes case?

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
 
-There are two distinct types of `BuildItem`, `SimpleBuildItem` and `MultiBuildItem`. `SimpleBuildItem` will only ever
-have a single instance created, while `MultiBuildItem` can have many instances.
+==== Build steps
```
Thanks, the doc is a massive step up!

However we need have some concrete examples so that people can read the concrete example and then read the abstract documentation and not be totally lost.
One simple, one multi,  one virtual build step and one @Weak usage based on real live example would go a long way.
I'd recommend to put these examples before and the detailed explanations after.

PS: I understand that you started not from the best guide and therefore more work falls on you by side effect. Sorry about that :)

dmlloyd replied:
It's no problem, I'll come up with something. :)

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
+Virtual build items are final (usually empty) classes which extend `io.quarkus.builder.item.VirtualBuildItem`.
+They cannot be instantiated, but represent "barriers" which can impose ordering between steps.  They can also be used in
+the same way that popular build systems use "pseudo-targets", which is to say that the build item can represent a
+conceptual goal that does not have a concrete representation.
```
Can you give an concrete example of when you need this barriers operation

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
+this is undesirable.
+
+For such cases, the `@io.quarkus.deployment.annotations.Weak` annotation may be used.  This annotation indicates that
+the build step should not automatically be included solely the basis of producing the annotated value.
```
s/solely the basis/solely on the basis/

emmanuelbernard commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
+of that build item.  Such a build step would use the `@Weak` annotation on the generated class build item, while normally
+producing the other build item.  If the other build item is ultimately consumed by something, then the step would run
+and the class would be generated.  If nothing consumes the other build item, the step would not be included in the build
+process.
```
I am totally lost on what is happening. Again I think something with a concrete example would help a lot to understand. Also what is this generated class? a bytecode generated one? And that one must be annotated with `@Weak`?

emmanuelbernard commented:
So probably I would replace `VirtualBuildItem` by `BarrierBuildItem` from what I understand. That's the only use case that is expressed at any rate.

dmlloyd commented:
> So probably I would replace `VirtualBuildItem` by `BarrierBuildItem` from what I understand. That's the only use case that is expressed at any rate.

There was a second use case given, which is to have a pseudo-target for build.  This is the use case that I am using right now, but I can illustrate it with some examples.

dmlloyd commented:
Looks like I mixed up a couple of references too.

emmanuelbernard commented:
> There was a second use case given, which is to have a pseudo-target for build. 

Ah right I had no idea what you meant. Still don't. As a life rule, I don't do build.

dmlloyd commented:
No worries, I'll come up with some example :)

dmlloyd commented:
OK I think I've addressed all of the issues.

emmanuelbernard commented:
Thanks @dmlloyd A couple of extra remarks:

- `@Consumes` `@Produces` will likely be named used by other frameworks. I think it mike make sense to qualify them e.g. `@ConsumesBuildItem` or `@BuildProduce` to align with `BuildProducer`
-  should `@Consumes*` / `@Produce*`  accept more than one BuildItem? Without changing the name (extra s) we could allow for a VBI[] value
- on the names have you explored @HappensBefore or something equivalent. I could see how it might be confusing though
- would it make sense to consume Single or even Multi build item?
- I did not get your pseudo-target example. looks like a barrier to me?
- thinking about the virtual name, would `@NamedBuildItem` or `@MarkedBuildItem` make sense?

On the weak annotation, I am finnally getting it if I think hard enough. Improvement proposal

```
Certain types of build items are always consumed, such as generated classes or resources.
An extension might produce a build item along with a generated class to facilitate the usage of that build item. The generated class only makes sense if the main build item is consumed. 
These side build steps (like the generated class in our example) would use the @Weak annotation on the generated class build item, while normally producing the other build item. If the other build item is ultimately consumed by something, then the step would run and the class would be generated. If nothing consumes the other build item, the step would not be included in the build process.

In our example, `GeneratedClassBuildItem` will be produced only if `ExecutorClassBuildItem` is consumed by another build step.
```

dmlloyd commented:
> Thanks @dmlloyd A couple of extra remarks:
> 
> * `@Consumes` `@Produces` will likely be named used by other frameworks. I think it mike make sense to qualify them e.g. `@ConsumesBuildItem` or `@BuildProduce` to align with `BuildProducer`

Hmm maybe.  I don't want to make them too unwieldy; maybe `@ConsumeItem` and `@ProduceItem`?

> * should `@Consumes*` / `@Produce*`  accept more than one BuildItem? Without changing the name (extra s) we could allow for a VBI[] value

They're already `@Repeatable`.

> * on the names have you explored @happensbefore or something equivalent. I could see how it might be confusing though

I did start off with them called `@Before`/`@After` but it's misleading because the step wouldn't come before/after the *item* (because that is nonsense),

> * would it make sense to consume Single or even Multi build item?

Yes, that does make sense.  I was going to make this change in a follow-up PR but I can put it in here instead, since I'm already probably making changes.

> * I did not get your pseudo-target example. looks like a barrier to me?

A barrier would be to make sure A runs before/after B.  A pseudo-target would be to cause a step to run that ordinarily wouldn't, and would represent something abstract (like to represent the fact that a native image, docker image, or uberjar was requested to be built).

> * thinking about the virtual name, would `@NamedBuildItem` or `@MarkedBuildItem` make sense?

I'm still pretty keen on "virtual" because it is accurate and pretty clear.

> On the weak annotation, I am finnally getting it if I think hard enough. Improvement proposal
> 
> ```
> Certain types of build items are always consumed, such as generated classes or resources.
> An extension might produce a build item along with a generated class to facilitate the usage of that build item. The generated class only makes sense if the main build item is consumed. 
> These side build steps (like the generated class in our example) would use the @Weak annotation on the generated class build item, while normally producing the other build item. If the other build item is ultimately consumed by something, then the step would run and the class would be generated. If nothing consumes the other build item, the step would not be included in the build process.
> 
> In our example, `GeneratedClassBuildItem` will be produced only if `ExecutorClassBuildItem` is consumed by another build step.
> ```

OK I can work this in there.

emmanuelbernard commented:
I think `@ConsumesItem` would be a bad compromise. So either just `@consumes` or `@ConsumesBuildItem`. My argument for the longer version is that this is not such a common feature as wedid not need it before so it can get a longer name.

emmanuelbernard commented:
One thing I don't get is that in the pseudo-target model, who will call this build item? Another build step right? with a `@Consumes*`. Which seems like a barrier to me.

dmlloyd commented:
> I think `@ConsumesItem` would be a bad compromise. So either just `@consumes` or `@ConsumesBuildItem`. My argument for the longer version is that this is not such a common feature as wedid not need it before so it can get a longer name.

How so?  Does its argument type not unambiguously define what is being consumed?  Keep in mind that the argument is always going to be a class whose name ends with `BuildItem` already.

> One thing I don't get is that in the pseudo-target model, who will call this build item? Another build step right? with a `@Consumes*`. Which seems like a barrier to me.

A pseudo-target would normally be added by the creator/mojo/whatever assembles the chain.

dmlloyd commented:
> * `@Consumes` `@Produces` will likely be named used by other frameworks. I think it mike make sense to qualify them e.g. `@ConsumesBuildItem` or `@BuildProduce` to align with `BuildProducer`

I was thinking about this some more and I don't think it is likely to matter.  It is already unlikely that someone would be producing and consuming virtual build items in the same class that they are also directly accessing classes called `Produce` or `Consume`.  If that unlikely event comes to pass, qualifying one or the other is clearly not the end of the world: it's a normal Java occurrence, nothing weird or scary about it.  At most this might affect one or two specific users, I would think.

I think I should just keep `Produce` and `Consume`.  And in a future update, `Require` is likely to appear, which would fit the same mold for the same reason.

emmanuelbernard commented:
@dmlloyd I am actually concerned about IDE auto completion, so you would get the Quarkus virtual build item Consume annotation every time you Cons + CTRL+SPACE

dmlloyd commented:
> @dmlloyd I am actually concerned about IDE auto completion, so you would get the Quarkus virtual build item Consume annotation every time you Cons + CTRL+SPACE

Build steps only occur within deployment modules, so I would expect again that this only would ever affect the few people who develop extensions that in turn deal with such frameworks.  The current number of such frameworks in Quarkus is zero: nothing else defines a class called `Produce` or `Consume` (or `Require` or `Weak` for that matter).

FroMage commented:
I know it's already a mess between JAX-RS's `@Produces` and CDI's `@Producer` (and same for consumer), so adding _another_ is bound to be confusing.

If I understand correctly, the virtual build items are like the current simple build items we use with no data, but they work with annotations rather than method signatures to declare dependencies.

For example:

```java
/**
 * Purely marker build item so that you can register enhancers after Hibernate
 * registers its enhancers, which would make your enhancers run before the
 * Hibernate enhancers
 */
public final class HibernateEnhancersRegisteredBuildItem extends SimpleBuildItem {

}

    @BuildStep
    public HibernateEnhancersRegisteredBuildItem enhancerDomainObjects(JpaEntitiesBuildItem domainObjects,
            BuildProducer<BytecodeTransformerBuildItem> transformers,
            List<AdditionalJpaModelBuildItem> additionalJpaModelBuildItems,
            BuildProducer<GeneratedClassBuildItem> additionalClasses) {
        // Modify the bytecode of all entities to enable lazy-loading, dirty checking, etc..
        enhanceEntities(domainObjects, transformers, additionalJpaModelBuildItems, additionalClasses);
        // this allows others to register their enhancers after Hibernate, so they run before ours
        return new HibernateEnhancersRegisteredBuildItem();
    }
```

The same would look like:

```java
/**
 * Purely marker build item so that you can register enhancers after Hibernate
 * registers its enhancers, which would make your enhancers run before the
 * Hibernate enhancers
 */
public final class HibernateEnhancersRegisteredBuildItem extends VirtualBuildItem {

}

    @BuildStep
    @Produce(HibernateEnhancersRegisteredBuildItem.class)
    public void enhancerDomainObjects(JpaEntitiesBuildItem domainObjects,
            BuildProducer<BytecodeTransformerBuildItem> transformers,
            List<AdditionalJpaModelBuildItem> additionalJpaModelBuildItems,
            BuildProducer<GeneratedClassBuildItem> additionalClasses) {
        // Modify the bytecode of all entities to enable lazy-loading, dirty checking, etc..
        enhanceEntities(domainObjects, transformers, additionalJpaModelBuildItems, additionalClasses);
    }
```

Then I'm not sure it worth it: we can already achieve that, in a way that is regular with other build items. Introducing a special way to do it (via annotations) that is not really less verbose can make it a little more confusing to learn, because the model is less straightforward.

I mean, yes we instantiate one less object, but the syntax is less regular, and not really much shorter.

Are you _sure_ it's worth it? We've been using hand-crafted virtual build items for a while and nobody asked for a better way to do it, unless I missed something?

dmlloyd commented:
There can be many producers and consumers for virtual items.  Isn't it a lot more natural to say `@Consume(Xxx.class)` than to inject an unused `List<Xxx>` that is filled with N empty Xxx instances?  Or at best to have to declare a singleton instance field for Xxx so that you can have a list of N copies of the same Xxx instance?

mkouba commented on core/deployment/src/main/java/io/quarkus/deployment/annotations/Produces.java:
```diff
@@ -0,0 +1,20 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface Produces {
```
Eh, pls no more`@Produces`... we have `javax.enterprise.inject.Produces` and `javax.ws.rs.Produces` already.

dmlloyd replied:
That's just the repeatable holder.  I've already renamed them to `ProduceRepeatable`, `ConsumeRepeatable`.

dmlloyd replied:
Not that it matters that much because this API is only visible to extension authors...

mkouba replied:
>  I've already renamed them to ProduceRepeatable, ConsumeRepeatable

+1

> Not that it matters that much because this API is only visible to extension authors...

Even the extension authors might get confused ;-). They will most likely see the CDI API and very often also JAX-RS API.

FroMage replied:
Hah, I told you this was a touchy subject ;)

FroMage replied:
Note that extension authors, when they work on the RESTEasy extension, for example, will already have the JAX-RS and CDI ones in their classpath because the deployment module imports the runtime modules. So yes, this is only visible to extension authors, but given that this PR feature is also only for them, we have to think about them with compassion and caring :)

FroMage commented:
> There can be many producers and consumers for virtual items. Isn't it a lot more natural to say @Consume(Xxx.class) than to inject an unused List<Xxx> that is filled with N empty Xxx instances? Or at best to have to declare a singleton instance field for Xxx so that you can have a list of N copies of the same Xxx instance?

OK, so that's something I didn't think of. 

ATM if you want to produce several virtual build items you add a `BuildProducer<MyBuildItem> buildItemProducer` per item. Contrary to the case where you produce a single one, the `@Consumes` case is indeed less typing for producing >1 item. For consuming >1 item, the same is not true, because we're rid of the `BuildProducer` extra token, so we're down to `MyBuildItem foo,` compared to an extra `MyBuildItem.class,` for the annotation. Still, it's a win for annotations for producing >1 item. Does that happen often? Is this often enough to warrant a model deviation just for virtual items? I don't have the answer, we should look at existing virtual build items and how common they are and how often they are produced by the same build step.

As for the `List` you mention, I'm not following. I'm assuming build items of separate classes, with different meanings, which you can't produce via `BuildProducer<List<?>>` or consume via `List<?>` because what would their common supertype be? How could we wire them in using static analysis?

Unless you're saying there's a use-case for producing the same virtual build item (same class) more than once? And consuming them all at once? I can't think of a use-case for that, so can you explain?

dmlloyd commented:
> > There can be many producers and consumers for virtual items. Isn't it a lot more natural to say @consume(Xxx.class) than to inject an unused List that is filled with N empty Xxx instances? Or at best to have to declare a singleton instance field for Xxx so that you can have a list of N copies of the same Xxx instance?
> 
> OK, so that's something I didn't think of.
> 
> ATM if you want to produce several virtual build items you add a `BuildProducer<MyBuildItem> buildItemProducer` per item. Contrary to the case where you produce a single one, the `@Consumes` case is indeed less typing for producing >1 item. For consuming >1 item, the same is not true, because we're rid of the `BuildProducer` extra token, so we're down to `MyBuildItem foo,` compared to an extra `MyBuildItem.class,` for the annotation. Still, it's a win for annotations for producing >1 item. Does that happen often? Is this often enough to warrant a model deviation just for virtual items? I don't have the answer, we should look at existing virtual build items and how common they are and how often they are produced by the same build step.

I'm not really sure that token count is the primary consideration.  If I'm implementing a build step, I want to have parameters or injections for the things I need, and possibly to return a thing I produce.  Ordering and pseudo-targets are strictly meta and belong with metadata.  The awkwardness of producing and consuming empty items is a side-effect of that.

> As for the `List` you mention, I'm not following. I'm assuming build items of separate classes, with different meanings, which you can't produce via `BuildProducer<List<?>>` or consume via `List<?>` because what would their common supertype be? How could we wire them in using static analysis?

If more than one thing can be before a given "barrier", then all of those things have to produce the build item.  If we're using empty classes for this purpose then they have to be multi and be consumed as a list.

To produce multi items you still use a plain `BuildProducer`, but to consume them you must consume a `List<TheItem>`.


> Unless you're saying there's a use-case for producing the same virtual build item (same class) more than once? And consuming them all at once? I can't think of a use-case for that, so can you explain?

Before logging starts, after logging starts.  Pretty much any before/after implies that more than one thing can come before the event, and more than one thing can come after it.

FroMage commented:
> Ordering and pseudo-targets are strictly meta and belong with metadata. The awkwardness of producing and consuming empty items is a side-effect of that

Except we can already implement them with the current API. It's generally bad design to have more than one way to do things (see Perl), and we already have two ways to do build steps (field injection or magic method injection). The fact that we can simplify this particular use-case by turning it into some annotations is of value, certainly, but has to be weighted against the regularity of the system and the fact that it introduces yet a third way to do things.

Don't get me wrong, I'm not opposed to this, I truely don't know if it's a good or bad thing. I'm just collecting what I see as advantages/drawbacks of both approaches so others can voice a stronger opinion.

At the very least, we'll have to strongly guide people towards not using empty build items anymore if this is merged, and we'll have to convert all existing empty build items into virtual ones.

> If more than one thing can be before a given "barrier", then all of those things have to produce the build item. If we're using empty classes for this purpose then they have to be multi and be consumed as a list.

OK, I didn't think of that use-case. Do we really have needs for those barriers ATM or are they hypotheticals? Do we currently have empty multi build items in our source code?

dmlloyd commented:
> > Ordering and pseudo-targets are strictly meta and belong with metadata. The awkwardness of producing and consuming empty items is a side-effect of that
> 
> Except we can already implement them with the current API. It's generally bad design to have more than one way to do things (see Perl), and we already have two ways to do build steps (field injection or magic method injection). The fact that we can simplify this particular use-case by turning it into some annotations is of value, certainly, but has to be weighted against the regularity of the system and the fact that it introduces yet a third way to do things.

I think the things that are being done are different enough to justify the multiple approaches.

> Don't get me wrong, I'm not opposed to this, I truely don't know if it's a good or bad thing. I'm just collecting what I see as advantages/drawbacks of both approaches so others can voice a stronger opinion.
> 
> At the very least, we'll have to strongly guide people towards not using empty build items anymore if this is merged, and we'll have to convert all existing empty build items into virtual ones.

Right.  I tend to see this as an improvement though.

> > If more than one thing can be before a given "barrier", then all of those things have to produce the build item. If we're using empty classes for this purpose then they have to be multi and be consumed as a list.
> 
> OK, I didn't think of that use-case. Do we really have needs for those barriers ATM or are they hypotheticals?

Yes we do.  This is how I expect we're going to implement extension-defined build targets (like the proposed native image extension).  This is also how I expect that user-defined startup handlers will be ordered, if that conversation ever resolves.

> Do we currently have empty multi build items in our source code?

I haven't checked.  If we do, they'd also be prime candidates for replacing with virtual items.

FroMage commented:
OK, in light of that, I suppose I'm in favour of this proposal.

dmlloyd commented:
I've updated the names of the repeating annotations.  I've also updated the `@Consume` annotation to allow any build item to be specified.

dmlloyd commented:
@emmanuelbernard do you have further feedback on this?

stuartwdouglas commented:
@dmlloyd this needs a rebase

dmlloyd commented:
Just waiting for CI.

dmlloyd commented:
@emmanuelbernard is there something else you wanted addressed?  I think we have a few things that could benefit from this change now.

dmlloyd committed:
Add a bit of extra clarifying verbiage

emmanuelbernard commented:
Didn't we settle on something else than Virtual as a name. Marker or something like that?

emmanuelbernard commented:
Ok so we never closed it.
There was `Virtual`, `Marker` and `Empty` as contenders. https://groups.google.com/forum/#!searchin/quarkus-dev/virtual$20marker%7Csort:date/quarkus-dev/h8hBo8DRV80/hf81OuabCQAJ

In that conversation you make two points about Marker:

1. Things should be named after what they are, not what they are helping do. I disagree and I think it showed into the `@Recorder` case. So let's consider this point aside.
2. Marker is only one use case, but I haven't found the other to see if marker was out of place.

Otherwise, between `Empty` and `Virtual`, I do prefer `Empty`
So I think we can go for `Marker` if the argument 2 is weak. Otherwise `Empty` which you were fine to compromise on.

dmlloyd commented:
I've updated the PR to use `Empty` instead of `Virtual`.

stuartwdouglas commented:
CI failed due to formatting issues

dmlloyd commented:
Ping @emmanuelbernard

dmlloyd commented:
Guillaume says master is open.  So...
